### PR TITLE
update_icon() improvements

### PR DESCRIPTION
**Pull request by MrPerson**
## About The Pull Request

Fixes some update_icon() calls to properly call parent and use update_overlays() and update_icon_state(). Also changes some code to not change state in update_icon() or handle behavior based on extra arguments into update_icon(). Both of these are wrong, don't do that.

Adds two elements: element/update_icon_blocker and element/update_icon_updates_onmob. The former just sends the signals to stop update_overlays() and update_icon_state() from being called, the latter calls update_inv_hands() or regenerate_icons() when update_icons() is called if that item is in a mob's hands or being worn by a mob.

## Why It's Good For The Game

Everything should call atom/update_icon(), it fires signals and does other good things we want to hook into but can't because of overrides like these.

### CODE REVIEW

File: code/game/atoms.dm [hunk 1]
```diff
@@ -505,6 +507,10 @@
 		if(length(new_overlays))
 			managed_overlays = new_overlays
 			add_overlay(new_overlays)
+		postSignal = TRUE
+
+	if(postSignal)
+		SEND_SIGNAL(src, COMSIG_ATOM_UPDATED_ICON)
```
Comment (ninjanomnom) [inline code/game/atoms.dm]:
```suggestion
		SEND_SIGNAL(src, COMSIG_ATOM_UPDATED_ICON, signalOut)
```
Also honestly you probably should call on this signal regardless of if changes happened

File: code/datums/elements/update_icon_blocker.dm [hunk 2]
```diff
@@ -0,0 +1,10 @@
+//Prevents calling anything in update_icon() like update_icon_state() or update_overlays()
+
+/datum/element/update_icon_blocker/Attach(atom/target)
+	if(!istype(target))
+		return ELEMENT_INCOMPATIBLE
+	RegisterSignal(target, COMSIG_ATOM_UPDATE_ICON, .proc/block_update_icon)
+	return ..()
```
Comment (ninjanomnom) [inline code/datums/elements/update_icon_blocker.dm]: prefer `. = ..()` at the top

File: code/datums/elements/update_icon_blocker.dm [hunk 3]
```diff
@@ -0,0 +1,10 @@
+//Prevents calling anything in update_icon() like update_icon_state() or update_overlays()
+
+/datum/element/update_icon_blocker/Attach(atom/target)
+	if(!istype(target))
```
Comment (ninjanomnom) [inline code/datums/elements/update_icon_blocker.dm]: I'm really not a fan of this pattern where the arg type may be incorrect and it's checked in the proc. Please keep the arg type correct for what could come in and cast it to another var if you need to istype it. In this case though you dont even need it cast properly

File: code/datums/elements/update_icon_updates_onmob.dm [hunk 4]
```diff
@@ -0,0 +1,15 @@
+//update_icon() may change the onmob icons
+//Very good name, I know
+
+/datum/element/update_icon_updates_onmob/Attach(obj/item/target)
```
Comment (ninjanomnom) [inline code/datums/elements/update_icon_updates_onmob.dm]: Ditto

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/game/machinery/iv_drip.dm [hunk 5]
```diff
@@ -240,14 +240,16 @@
...
-    . = ..()
-    beaker = new /obj/item/reagent_containers/glass/saline(src)
+	. = ..()
+	beaker = new /obj/item/reagent_containers/glass/saline(src)
 
-/obj/machinery/iv_drip/saline/update_icon()
-    return
+/obj/machinery/iv_drip/saline/ComponentInitialize()
+	AddElement(/datum/element/update_icon_blocker)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/machinery/iv_drip.dm]: `. = ..()` on top

File: code/datums/components/armor_plate.dm [hunk 6]
```diff
@@ -12,6 +12,7 @@
 	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examine)
 	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/applyplate)
 	RegisterSignal(parent, COMSIG_PARENT_PREQDELETED, .proc/dropplates)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/apply_overlays)
```
Comment (ninjanomnom) [inline code/datums/components/armor_plate.dm]: Check if it's a ripley before registering and you don't need to istype in the receiving proc

File: code/game/objects/items/RCL.dm [hunk 7]
```diff
@@ -23,6 +23,10 @@
 	var/datum/radial_menu/persistent/wiring_gui_menu
 	var/mob/listeningTo
 
+/obj/item/twohanded/rcl/ComponentInitialize()
+	AddElement(/datum/element/update_icon_updates_onmob)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/RCL.dm]: `. = ..()` on top

File: code/game/objects/items/chrono_eraser.dm [hunk 8]
```diff
@@ -65,8 +65,9 @@
 		TED = new(src.loc)
 		return INITIALIZE_HINT_QDEL
 
-/obj/item/gun/energy/chrono_gun/update_icon()
-	return
+/obj/item/gun/energy/chrono_gun/ComponentInitialize()
+	AddElement(/datum/element/update_icon_blocker)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/chrono_eraser.dm]: `. = ..()` on top

File: code/game/objects/items/defib.dm [hunk 9]
```diff
@@ -314,6 +309,9 @@
 	var/tlimit = DEFIB_TIME_LIMIT * 10
 	var/base_icon_state = "defibpaddles"
 
+/obj/item/twohanded/shockpaddles/ComponentInitialize()
+	AddElement(/datum/element/update_icon_updates_onmob)
```
Comment (ninjanomnom) [inline code/game/objects/items/defib.dm]: Parent call?

File: code/game/objects/items/dice.dm [hunk 10]
```diff
@@ -153,8 +153,9 @@ obj/item/dice/d6/ebony
 	w_class = WEIGHT_CLASS_SMALL
 	sides = 100
 
-/obj/item/dice/d100/update_icon()
-	return
+/obj/item/dice/d100/ComponentInitialize()
+	AddElement(/datum/element/update_icon_blocker)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/dice.dm]: `. = ..()` on top

File: code/game/objects/items/dice.dm [hunk 11]
```diff
@@ -163,8 +164,9 @@ obj/item/dice/d6/ebony
 	sides = 20
 	special_faces = list("It is certain","It is decidedly so","Without a doubt","Yes, definitely","You may rely on it","As I see it, yes","Most likely","Outlook good","Yes","Signs point to yes","Reply hazy try again","Ask again later","Better not tell you now","Cannot predict now","Concentrate and ask again","Don't count on it","My reply is no","My sources say no","Outlook not so good","Very doubtful")
 
-/obj/item/dice/eightbd20/update_icon()
-	return
+/obj/item/dice/eightbd20/ComponentInitialize()
+	AddElement(/datum/element/update_icon_blocker)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/dice.dm]: `. = ..()` on top

File: code/game/objects/items/dice.dm [hunk 12]
```diff
@@ -173,8 +175,9 @@ obj/item/dice/d6/ebony
 	sides = 48
 	special_faces = list("Cube-Side: 1-1","Cube-Side: 1-2","Cube-Side: 1-3","Cube-Side: 1-4","Cube-Side: 1-5","Cube-Side: 1-6","Cube-Side: 2-1","Cube-Side: 2-2","Cube-Side: 2-3","Cube-Side: 2-4","Cube-Side: 2-5","Cube-Side: 2-6","Cube-Side: 3-1","Cube-Side: 3-2","Cube-Side: 3-3","Cube-Side: 3-4","Cube-Side: 3-5","Cube-Side: 3-6","Cube-Side: 4-1","Cube-Side: 4-2","Cube-Side: 4-3","Cube-Side: 4-4","Cube-Side: 4-5","Cube-Side: 4-6","Cube-Side: 5-1","Cube-Side: 5-2","Cube-Side: 5-3","Cube-Side: 5-4","Cube-Side: 5-5","Cube-Side: 5-6","Cube-Side: 6-1","Cube-Side: 6-2","Cube-Side: 6-3","Cube-Side: 6-4","Cube-Side: 6-5","Cube-Side: 6-6","Cube-Side: 7-1","Cube-Side: 7-2","Cube-Side: 7-3","Cube-Side: 7-4","Cube-Side: 7-5","Cube-Side: 7-6","Cube-Side: 8-1","Cube-Side: 8-2","Cube-Side: 8-3","Cube-Side: 8-4","Cube-Side: 8-5","Cube-Side: 8-6")
 
-/obj/item/dice/fourdd6/update_icon()
-	return
+/obj/item/dice/fourdd6/ComponentInitialize()
+	AddElement(/datum/element/update_icon_blocker)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/dice.dm]: `. = ..()` on top

File: code/game/objects/items/flamethrower.dm [hunk 13]
```diff
@@ -29,6 +29,10 @@
 	var/igniter_type = /obj/item/assembly/igniter
 	trigger_guard = TRIGGER_GUARD_NORMAL
 
+/obj/item/flamethrower/ComponentInitialize()
+	AddElement(/datum/element/update_icon_updates_onmob)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/flamethrower.dm]: `. = ..()` on top

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 14]
```diff
@@ -305,24 +305,27 @@
...
 	else
 		icon_state = "cyborg_upgrade5"
 
 /obj/item/borg/upgrade/selfrepair/proc/activate_sr()
 	START_PROCESSING(SSobj, src)
 	on = TRUE
 	update_icon()
+	for(var/X in actions)
+		var/datum/action/A = X
+		A.UpdateButtonIcon()
```
Comment (ninjanomnom) [inline code/game/objects/items/robot/robot_upgrades.dm]: You made a signal for this sort of thing

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 15]
```diff
@@ -305,24 +305,27 @@
...
 
 /obj/item/borg/upgrade/selfrepair/proc/deactivate_sr()
 	STOP_PROCESSING(SSobj, src)
 	on = FALSE
 	update_icon()
+	for(var/X in actions)
+		var/datum/action/A = X
+		A.UpdateButtonIcon()
```
Comment (ninjanomnom) [inline code/game/objects/items/robot/robot_upgrades.dm]: Ditto

File: code/game/objects/items/shields.dm [hunk 16]
```diff
@@ -130,6 +130,10 @@
 	. = ..()
 	embedded_flash = new(src)
 
+/obj/item/shield/riot/flash/ComponentInitialize()
+	AddElement(/datum/element/update_icon_updates_onmob)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/shields.dm]: `. = ..()` on top

File: code/game/objects/items/stacks/medical.dm [hunk 17]
```diff
@@ -239,13 +239,14 @@
...
 
 
-/obj/item/stack/medical/mesh/update_icon()
+/obj/item/stack/medical/mesh/update_icon_state()
 	if(!is_open)
-		return
-	. = ..()
+		icon_state = "regen_mesh_closed"
+	else
+		..()
```
Comment (ninjanomnom) [inline code/game/objects/items/stacks/medical.dm]:
```suggestion
		return ..()
```

File: code/game/objects/items/stacks/rods.dm [hunk 18]
```diff
@@ -77,8 +77,9 @@ GLOBAL_LIST_INIT(rod_recipes, list ( \
 	is_cyborg = 1
 	cost = 250
 
-/obj/item/stack/rods/cyborg/update_icon()
-	return
+/obj/item/stack/rods/cyborg/ComponentInitialize()
+	AddElement(/datum/element/update_icon_blocker)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/stacks/rods.dm]: `. = ..()` on top

File: code/game/objects/items/tools/weldingtool.dm [hunk 19]
```diff
@@ -42,23 +42,25 @@
 	reagents.add_reagent(/datum/reagent/fuel, max_fuel)
 	update_icon()
 
+/obj/item/weldingtool/ComponentInitialize()
+	AddElement(/datum/element/update_icon_updates_onmob)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/tools/weldingtool.dm]: `. = ..()` on top

File: code/game/objects/items/toys.dm [hunk 20]
```diff
@@ -47,6 +47,10 @@
 	. = ..()
 	create_reagents(10)
 
+/obj/item/toy/waterballoon/ComponentInitialize()
+	AddElement(/datum/element/update_icon_updates_onmob)
+	return ..()
```
Comment (ninjanomnom) [inline code/game/objects/items/toys.dm]: `. = ..()` on top

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/game/objects/items/robot/robot_upgrades.dm (hunk 14 above)
Comment (MrPerson) [reply code/game/objects/items/robot/robot_upgrades.dm]: I'm gonna have to do this carefully but sure, I'll get around to it

Comment (MrPerson) [reply code/game/objects/items/robot/robot_upgrades.dm]: Actually I'm not sure what to do here, have the action button hook into the updated icon signal to call UpdateButtonIcon()?

Comment (ninjanomnom) [reply code/game/objects/items/robot/robot_upgrades.dm]: yeah exactly

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**